sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala

```diff
@@ -174,6 +174,23 @@ object DecimalPrecision extends TypeCoercionRule {
       CheckOverflow(Pmod(promotePrecision(e1, widerType), promotePrecision(e2, widerType)),
         resultType, nullOnOverflow)
 
+    case expr @ IntegralDivide(
+        e1 @ DecimalType.Expression(p1, s1), e2 @ DecimalType.Expression(p2, s2)) =>
+      val widerType = widerDecimalType(p1, s1, p2, s2)
+      val promotedExpr =
+        IntegralDivide(promotePrecision(e1, widerType), promotePrecision(e2, widerType))
+      if (expr.dataType.isInstanceOf[DecimalType]) {
+        // This follows division rule
+        val intDig = p1 - s1 + s2
+        // No precision loss can happen as the result scale is 0.
```

Comment:
I don't quite get it. How can we avoid overflow for division? e.g. `div(999..., 0.000...1)`, we can't represent the result with decimal.

Reply:
Because in a case like you mentioned, the overflow doesn't happen in the `CheckOverflow`, but it happens in the `PromotePrecision` on the operands. If no overflows happen on the operands, since the result has scale 0, we cannot have more than 38 digits of the integer part and no overflow can happen.

I mean, either there is an overflow in the promote precision of the operands or no overflow can happen at all.

Reply:
do we have a test case for operand overflow?

Reply:
no, I am adding it, thanks.